### ISSUE TITLE
integer array overflows with 3km and higher resolutions in EAM

### ISSUE THREAD

**Issue by mt5555** [E3SM-Project/E3SM#2698]
EAM has the same problem as standalone HOMME, issue #2697 
For 3km at larger grids (NE=1024), the following fixes will be necessary (but may not be sufficient, as EAM may have other issues as well at these resolutions)

1) For all calls to pio_initdecomp(), the "dof" array needs to be integer\*8 (PIO_OFFSET_KIND)
The computations for dof() need to be done as integer\*8, although the individual variables used in the computations can remain integer\*4.  in HOMME, I fixed this by converting them with INT(var,8)  

2) For NE>=6144 (500m), the restart will be broken, as the "dims" array will also need to be updated to integer\*8.  However, it appears that PIO1 and PIO2 does not have a pio_initdecomp() function that supports integer\*8  dims array.   

It's unlikely that we will run NE=6144 anytime soon, so the second issue is a low priority.

**Comment by mrnorman** [comment]
The physics will also overflow, and I believe I've fixed it with:

```
diff --git a/components/cam/src/utils/cam_grid_support.F90 b/components/cam/src/utils/cam_grid_support.F90
index d0143c3..a25807b 100644
--- a/components/cam/src/utils/cam_grid_support.F90
+++ b/components/cam/src/utils/cam_grid_support.F90
@@ -2559,7 +2559,7 @@ contains
     ! Local variables
     integer                                :: dims(2)
     integer                                :: dstrt, dend
-    integer                                :: gridlen, gridloc, ierr
+    integer(iMap)                          :: gridlen, gridloc, ierr
 
     ! Check to make sure the map meets our needs
     call this%coord_lengths(dims)
@@ -2582,7 +2582,7 @@ contains
       gridloc = count((map(dstrt,:) /= 0) .and. (map(dend,:) /= 0))
     end if
     call MPI_Allreduce(gridloc, gridlen, 1, MPI_INTEGER, MPI_SUM, mpicom, ierr)
-    if (gridlen /= product(dims)) then
+    if (gridlen /= product(int(dims,iMap))) then
       call endrun('cam_grid_set_map: Bad map size for '//trim(this%name))
     else
       if (.not. associated(this%map)) then
diff --git a/components/cam/src/utils/cam_map_utils.F90 b/components/cam/src/utils/cam_map_utils.F90
index 4e45c88..4bdfcd3 100644
--- a/components/cam/src/utils/cam_map_utils.F90
+++ b/components/cam/src/utils/cam_map_utils.F90
@@ -683,7 +683,7 @@ contains
     end if
 
     !
-    fileSize = product(filelens)
+    fileSize = product(int(filelens,iMap))
     srccnt = size(fieldlens)
     srclens(1:srccnt) = fieldlens(1:srccnt)
     if (srccnt < 7) then
@@ -691,7 +691,7 @@ contains
     end if
 
     ! Allocate output filemap (dof)
-    allocate(filemap(product(fieldlens)))
+    allocate(filemap(product(int(fieldlens,iMap))))
     filemap = 0
 
     ! Find map source dimensions in input array

```

**Comment by mt5555** [comment]
adding link to Matt's PR for the physics restart fix in the ECP fork:

https://github.com/E3SM-Project/ACME-ECP/pull/77

**Comment by brhillman** [comment]
This was enough to fix integer overflow with ne512, but not for ne1024.

**Comment by mt5555** [comment]
closing, and making a new issue for the one aspect of this issue not yet fixed:  #4763
